### PR TITLE
Return cache config as response even though found and created cache config could not put into cache configs inside cache service

### LINKED ISSUE #7208: regression cacheManager.getCache(“defined-in-xml")); can rarely return null
I think we have a regression in     

cacheManager.getCache(“defined-in-xml"));   

it looks like some times it returns null

i think its very rare.  I saw it happen when starting the first round of simulator tests for 3.6-EA3

and again just now running some other  j cache  tests
the latest one was

http://54.87.52.100/~jenkins/workspace/hd-jcache-nearCache/3.6-SNAPSHOT/2015_12_29-17_10_41/output/2015_12_29-17_10_59/failures-2015-12-29__17_11_00.txt

```
Failure[
   message='Worked ran into an unhandled exception'
   type=WORKER_EXCEPTION
   timestamp=1451401890070
   workerAddress=C_A1_W5
   agentAddress=52.90.126.123
   hzAddress=client:52.90.126.123
   workerId=worker-52.90.126.123-5-client
   test=TestCase{
        id=icache-sim,
        class=com.hazelcast.enterprise.tests.jcache.JcacheSim,
        asyncWait=50,
        cacheBaseNamesStr=nat-NoNear,
        cachesPerName=4,
        getAsyncThreds=4,
        getkeyDomainMax=1000000,
        loadCachePhase=true,
        loadkeyDomainMax=1000000,
        multiCacheHitCount=3,
        putAsyncThreds=16,
        putkeyDomainMax=1500000
    }
   cause=java.lang.NullPointerException
	at com.hazelcast.enterprise.tests.jcache.JcacheSim.validateCaceContent(JcacheSim.java:132)
	at com.hazelcast.enterprise.tests.jcache.JcacheSim.setup(JcacheSim.java:122)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.hazelcast.simulator.utils.ReflectionUtils.invokeMethod(ReflectionUtils.java:105)
	at com.hazelcast.simulator.worker.TestContainer.invoke(TestContainer.java:157)
	at com.hazelcast.simulator.protocol.processors.TestOperationProcessor$1.doRun(TestOperationProcessor.java:115)
	at com.hazelcast.simulator.protocol.processors.TestOperationProcessor$OperationThread.run(TestOperationProcessor.java:182)
]
```

https://github.com/hazelcast/hazelcast-ee-simulator-tests/blob/master/ee-tests/src/main/java/com/hazelcast/enterprise/tests/jcache/JcacheSim.java#L132

I have run this test lots of time,  and at random we hit the NPE on using the cache

From 3.6-EA3

```
Failure[
   message='Worked ran into an unhandled exception'
   type=WORKER_EXCEPTION
   timestamp=1450455278435
   workerAddress=C_A1_W4
   agentAddress=54.165.157.166
   hzAddress=client:54.165.157.166
   workerId=worker-54.165.157.166-4-client
   test=TestCase{
        id=icacheMaxSmall,
        class=com.hazelcast.simulator.tests.icache.EvictionICacheTest,
        basename=maxCachSmall1
    }
   cause=java.lang.NullPointerException
	at com.hazelcast.simulator.tests.icache.EvictionICacheTest.setup(EvictionICacheTest.java:95)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.hazelcast.simulator.utils.ReflectionUtils.invokeMethod(ReflectionUtils.java:105)
	at com.hazelcast.simulator.worker.TestContainer.invoke(TestContainer.java:157)
	at com.hazelcast.simulator.protocol.processors.TestOperationProcessor$1.doRun(TestOperationProcessor.java:115)
	at com.hazelcast.simulator.protocol.processors.TestOperationProcessor$OperationThread.run(TestOperationProcessor.java:182)
]
```

https://github.com/hazelcast/hazelcast-simulator/blob/80b8b1c6a43f5b9900551408da8bce20a9076ec3/tests/src/main/java/com/hazelcast/simulator/tests/icache/EvictionICacheTest.java#L95

### PR DESCRIPTION
**Pull request by serkan-ozal**
… in case of parallel access

Fixes #7208

### CODE REVIEW

**Comment by serkan-ozal** [comment]
This PR will also be applied to 3.6-RC1 and maintenance-3.x branches

**Comment by serkan-ozal** [comment]
run-lab-run

**Comment by emrahkocaman** [comment]
:+1:

**Comment by serkan-ozal** [comment]
Unrelated failure:

```
14:28:04 Tests run: 6, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 308.328 sec <<< FAILURE! - in com.hazelcast.concurrent.idgen.IdGeneratorBasicDistributedTest
14:28:04 testInit(com.hazelcast.concurrent.idgen.IdGeneratorBasicDistributedTest)  Time elapsed: 301.176 sec  <<< ERROR!
14:28:04 org.junit.runners.model.TestTimedOutException: test timed out after 300000 milliseconds
14:28:04 	at java.lang.Object.wait(Native Method)
14:28:04 	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.pollResponse(InvocationFuture.java:300)
14:28:04 	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.waitForResponse(InvocationFuture.java:245)
14:28:04 	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:222)
14:28:04 	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:202)
14:28:04 	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.getSafely(InvocationFuture.java:214)
14:28:04 	at com.hazelcast.concurrent.atomiclong.AtomicLongProxy.getAndIncrement(AtomicLongProxy.java:164)
14:28:04 	at com.hazelcast.concurrent.idgen.IdGeneratorProxy.newId(IdGeneratorProxy.java:73)
14:28:04 	at com.hazelcast.concurrent.idgen.IdGeneratorAbstractTest.testInit(IdGeneratorAbstractTest.java:48)
14:28:04 	at com.hazelcast.concurrent.idgen.IdGeneratorAbstractTest.testInit(IdGeneratorAbstractTest.java:36)
14:28:04 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
14:28:04 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
14:28:04 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
14:28:04 	at java.lang.reflect.Method.invoke(Method.java:597)
14:28:04 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
14:28:04 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
14:28:04 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
14:28:04 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
14:28:04 	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
14:28:04 	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
14:28:04 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
14:28:04 	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
14:28:04 	at java.lang.Thread.run(Thread.java:662)
```

File: hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetConfigOperation.java [hunk 1]
```diff
@@ -50,18 +50,15 @@ public CacheGetConfigOperation(String name, String simpleName) {
...
-                    cacheConfigFromSimpleConfig.setName(simpleName);
-                    cacheConfigFromSimpleConfig.setManagerPrefix(name.substring(0, name.lastIndexOf(simpleName)));
-                    if (service.putCacheConfigIfAbsent(cacheConfigFromSimpleConfig) == null) {
-                        response = cacheConfigFromSimpleConfig;
-                        return;
-                    }
+                    cacheConfig = new CacheConfig(simpleConfig);
+                    cacheConfig.setName(simpleName);
+                    cacheConfig.setManagerPrefix(name.substring(0, name.lastIndexOf(simpleName)));
+                    service.putCacheConfigIfAbsent(cacheConfig);
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetConfigOperation.java]: You assume the existing config will always be exactly the same as the one you just created from the simple cache config. Is this assumption always correct?

**Comment by emrahkocaman** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetConfigOperation.java (hunk 1 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetConfigOperation.java]: I have thought again and I think it is better to return existing one (not created one) if "putIfAbsent" fails

Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetConfigOperation.java]: excellent, it is the safest thing to do.

Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetConfigOperation.java]: Jaromir in fact they can be mutated. For example entry listener config. As I have mentioned above, ı will return existing one if there is

Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetConfigOperation.java]: :+1:

**Comment by serkan-ozal** [comment]
@emrahkocaman @jerrinot I have updated the PR

**Comment by jerrinot** [comment]
:+1: on green
well done!

**Comment by emrahkocaman** [comment]
:+1:

**Pull request merged**